service/storage_proxy.cc

```diff
@@ -3004,10 +3004,10 @@ future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mut
     }, std::forward<Range>(mutations), cl, type, std::move(permit), std::move(create_handler));
 }
 
-template<typename Range>
-future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mutate_prepare(Range&& mutations, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit) {
-    return mutate_prepare<>(std::forward<Range>(mutations), cl, type, std::move(permit), [this, tr_state = std::move(tr_state), allow_limit] (const typename std::decay_t<Range>::value_type& m, db::consistency_level cl, db::write_type type, service_permit permit) mutable {
-        return create_write_response_handler(m, cl, type, tr_state, std::move(permit), allow_limit);
+template<typename Range, typename... Args>
+future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mutate_prepare(Range&& mutations, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit, Args&&... args) {
+    return mutate_prepare<>(std::forward<Range>(mutations), cl, type, std::move(permit), [this, tr_state = std::move(tr_state), allow_limit, ... args = std::forward<Args>(args)] (const typename std::decay_t<Range>::value_type& m, db::consistency_level cl, db::write_type type, service_permit permit) mutable {
+        return create_write_response_handler(m, cl, type, tr_state, std::move(permit), allow_limit, std::forward<Args>(args)...);
```

Comment:
We should reduce template bloat. If mutate_prepare always accepts an erm, we don't need this.

Reply:
btw, we should pass std::span<> instead of Range, if the callers permit.

Reply:
mutate_prepare doesn't always accepts an erm. mutate_prepare is effectively a wrapper template call on top of create_write_response_handler for different "range" types. There are 5 create_write_response_handler overloads and only one needs erm

Reply:
(having 5 different create_write_response_handler:s doing slightly different things is an abuse of overloading; overloading should be used when they do exactly the same thing)

Reply:
But I see this:

```c++
template<typename Range, typename CreateWriteHandler>
future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mutate_prepare(Range&& mutations, db::consistency_level cl, db::write_type type, service_permit permit, CreateWriteHandler create_handler) {
    // apply is used to convert exceptions to exceptional future
    return futurize_invoke([this] (Range&& mutations, db::consistency_level cl, db::write_type type, service_permit permit, CreateWriteHandler create_handler) {
        unique_response_handler_vector ids;
        ids.reserve(std::distance(std::begin(mutations), std::end(mutations)));
        for (auto& m : mutations) {
            auto r_handler = create_handler(m, cl, type, permit);
            if (!r_handler) {
                return make_ready_future<result<unique_response_handler_vector>>(std::move(r_handler).as_failure());
            }
            ids.emplace_back(*this, std::move(r_handler).value());
        }
        return make_ready_future<result<unique_response_handler_vector>>(std::move(ids));
    }, std::forward<Range>(mutations), cl, type, std::move(permit), std::move(create_handler));
}
```

Here it's used to make iterating over mutations orthogonal from creating the response handler, which seems reasonable.

Reply:
Could well have used std::span and a function pointer.

Reply:
> But I see this:
> 
> ```c++
> template<typename Range, typename CreateWriteHandler>
> future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mutate_prepare(Range&& mutations, db::consistency_level cl, db::write_type type, service_permit permit, CreateWriteHandler create_handler) {
>     // apply is used to convert exceptions to exceptional future
>    ...
> }
> ```
> 
> Here it's used to make iterating over mutations orthogonal from creating the response handler, which seems reasonable.

Yes, but this one is not called by anyone but (!) another mutate_prepare() which's right below that one and which I meant to be a wrapper over create_write_response_handler()

Reply:
> > But I see this:
> > ```c++
> > template<typename Range, typename CreateWriteHandler>
> > future<result<storage_proxy::unique_response_handler_vector>> storage_proxy::mutate_prepare(Range&& mutations, db::consistency_level cl, db::write_type type, service_permit permit, CreateWriteHandler create_handler) {
> >     // apply is used to convert exceptions to exceptional future
> >    ...
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here it's used to make iterating over mutations orthogonal from creating the response handler, which seems reasonable.
> 
> Yes, but this one is not called by anyone but (!) another mutate_prepare() which's right below that one and which I meant to be a wrapper over create_write_response_handler()

I see that it at least called in `send_batchlog_mutation` as well.

Reply:
Yes, technically it is.